File: backend/src/deaths_door/game_manager.py
"""Game manager with event sourcing dispatch pattern."""

from __future__ import annotations

import asyncio
from datetime import datetime, timezone
from uuid import UUID, uuid4

from .apply import apply, replay
from .event_store import EventStore
from .events import (
    DeadVoteUsedSet,
    EventPayload,
    GameCreated,
    GameEvent,
    PlayerAdded,
    PlayerAliveSet,
    RolesIncluded,
    StatusEffectAdded,
)
from .game_state import GameState


class GameManager:
    """
    Manages game state via event sourcing.

    All mutations go through dispatch(), which atomically:
    1. Creates an event from the payload
    2. Applies it to produce a new state (validates)
    3. Persists the event to SQLite
    4. Updates the in-memory cache
    """

    def __init__(self, event_store: EventStore) -> None:
        """Create a new game manager backed by an event store."""
        self._store = event_store
        self._state: GameState | None = None
        self._lock = asyncio.Lock()

    @property
    def state(self) -> GameState:
        """Get the current game state. Raises if no game is active."""
        if self._state is None:
            raise RuntimeError("No active game")
        return self._state

    async def dispatch(self, payload: EventPayload) -> GameState:
        """Atomically apply and persist an event. Returns the new state."""
        async with self._lock:
            current = self.state
            event = GameEvent(
                game_id=current.game_id,
                sequence=current.version,
                timestamp=datetime.now(timezone.utc),
                payload=payload,
            )
            # Apply first to validate (raises on invalid state transitions)
            new_state = apply(current, event)
            # Persist only after successful apply
            self._store.append(event)
            self._state = new_state
            return new_state

    async def get_state(self) -> GameState:
        """Get the current game state, auto-creating if none exists."""
        if self._state is None:
            # Try loading the most recent game from the store
            game_ids = self._store.get_all_game_ids()
            if game_ids:
                await self.load_game(game_ids[-1])
            else:
                await self._create_sample_game()
        return self.state

    async def _create_sample_game(self) -> None:
        """Create a sample game for development/testing."""
        await self.create_game("trouble_brewing")

        roles = [
            "Imp", "Baron", "Poisoner", "Scarlet Woman",  # evil
            "Recluse", "Librarian", "Empath", "Investigator",  # good
            "Mayor", "Fortune Teller", "Slayer", "Monk", "Virgin",  # good
        ]
        await self.dispatch(RolesIncluded(names=tuple(roles)))

        # Named players with specific roles
        players = [
            ("Ryan", "Baron", "evil"),
            ("Yash", "Virgin", "good"),
            ("Other Ryan", "Imp", "evil"),
            ("Other Yash", "Poisoner", "evil"),
            ("Yet Another Ryan", "Scarlet Woman", "evil"),
            ("Yet Another Yash", "Recluse", "good"),
            ("Even More Ryan", "Librarian", "good"),
            ("Even More Yash", "Empath", "good"),
            ("Even Even More Ryan", "Mayor", "good"),
            ("Even Even More Yash", "Fortune Teller", "good"),
            ("Claude", "Monk", "good"),
        ]
        for name, role, alignment in players:
            await self.dispatch(PlayerAdded(player_name=name, character_name=role, alignment=alignment))

        # Status effects
        await self.dispatch(StatusEffectAdded(player_name="Yash", effect="Drunk"))
        await self.dispatch(StatusEffectAdded(player_name="Yash", effect="No Ability"))
        await self.dispatch(StatusEffectAdded(player_name="Ryan", effect="Is The Demon"))

        # Yash is dead and has used their vote
        await self.dispatch(PlayerAliveSet(player_name="Yash", is_alive=False))
        await self.dispatch(DeadVoteUsedSet(player_name="Yash", has_used_dead_vote=True))

    async def create_game(self, script_name: str) -> GameState:
        """Create a new game with a fresh game_id."""
        async with self._lock:
            game_id = uuid4()
            payload = GameCreated(script_name=script_name)
            event = GameEvent(
                game_id=game_id,
                sequence=0,
                timestamp=datetime.now(timezone.utc),
                payload=payload,
            )
            initial = GameState(game_id=game_id, script_name="")
            new_state = apply(initial, event)
            self._store.append(event)
            self._state = new_state
            return new_state

    async def load_game(self, game_id: UUID) -> GameState:
        """Load a game from the event store by replaying its events."""
        async with self._lock:
            events = self._store.get_events(game_id)
            if not events:
                raise ValueError(f"No events found for game {game_id}")
            self._state = replay(events)
            return self._state

    async def rewind(self, to_version: int) -> GameState:
        """
        Rewind the current game to a specific version.

        Deletes events after the target version and rebuilds state.
        """
        async with self._lock:
            current = self.state
            if to_version < 1 or to_version > current.version:
                raise ValueError(f"Invalid version {to_version} (current: {current.version})")

            # Delete events after the target and rebuild
            self._store.delete_after_sequence(current.game_id, to_version - 1)
            events = self._store.get_events(current.game_id)
            self._state = replay(events)
            return self._state

    async def fork(self, from_version: int) -> GameState:
        """
        Fork the current game from a specific version.

        Creates a new game_id with events copied up to from_version,
        then loads and returns the forked game.
        """
        async with self._lock:
            current = self.state
            if from_version < 1 or from_version > current.version:
                raise ValueError(f"Invalid version {from_version} (current: {current.version})")

            new_game_id = self._store.fork_game(current.game_id, up_to_sequence=from_version - 1)
            events = self._store.get_events(new_game_id)
            self._state = replay(events)
            return self._state

    async def get_history(self) -> list[GameEvent]:
        """Get all events for the current game."""
        current = self.state
        return self._store.get_events(current.game_id)

    async def list_games(self) -> list[UUID]:
        """List all game IDs in the store."""
        return self._store.get_all_game_ids()

    def has_active_game(self) -> bool:
        """Check if there is an active game loaded."""
        return self._state is not None


# Module-level singleton
event_store = EventStore()
game_manager = GameManager(event_store)

File: backend/src/game_manager.rs
//! Game manager with event-sourcing dispatch pattern.
//!
//! All mutations go through [`GameManager::dispatch`], which atomically:
//! 1. Creates an event from the payload,
//! 2. Applies it to produce a new state (validating),
//! 3. Persists the event to SQLite,
//! 4. Updates the in-memory cache and notifies SSE subscribers.

use chrono::Utc;
use tokio::sync::{broadcast, Mutex};
use uuid::Uuid;

use crate::apply::{apply, replay};
use crate::error::GameError;
use crate::event_store::EventStore;
use crate::events::{EventPayload, GameEvent};
use crate::game_state::GameState;

struct Inner {
    store: EventStore,
    state: Option<GameState>,
}

pub struct GameManager {
    inner: Mutex<Inner>,
    tx: broadcast::Sender<GameState>,
    /// Serializes lazy initialization in `get_state` so concurrent first
    /// requests (overlay SSE, iOS, timer) don't each create/load a game.
    init_lock: Mutex<()>,
}

impl GameManager {
    pub fn new(store: EventStore) -> Self {
        let (tx, _rx) = broadcast::channel(128);
        GameManager {
            inner: Mutex::new(Inner { store, state: None }),
            tx,
            init_lock: Mutex::new(()),
        }
    }

    /// Subscribe to state changes (used by the SSE stream).
    pub fn subscribe(&self) -> broadcast::Receiver<GameState> {
        self.tx.subscribe()
    }

    /// Re-broadcast the current state to SSE subscribers without a mutation.
    /// Used by the timer to push per-second updates over the same stream.
    pub async fn notify_current(&self) {
        if let Some(state) = self.inner.lock().await.state.clone() {
            let _ = self.tx.send(state);
        }
    }

    /// Get the current game state, or an error if no game is active.
    pub async fn state(&self) -> Result<GameState, GameError> {
        self.inner
            .lock()
            .await
            .state
            .clone()
            .ok_or(GameError::NoActiveGame)
    }

    pub async fn has_active_game(&self) -> bool {
        self.inner.lock().await.state.is_some()
    }

    /// Atomically apply and persist an event. Returns the new state.
    pub async fn dispatch(&self, payload: EventPayload) -> Result<GameState, GameError> {
        let mut inner = self.inner.lock().await;
        let current = inner.state.clone().ok_or(GameError::NoActiveGame)?;
        let event = GameEvent::new(current.game_id, current.version, payload);
        // Apply first to validate, then persist only after a successful apply.
        let new_state = apply(&current, &event);
        inner.store.append(&event)?;
        inner.state = Some(new_state.clone());
        // Non-blocking broadcast; ignored if there are no subscribers.
        let _ = self.tx.send(new_state.clone());
        Ok(new_state)
    }

    /// Get the current game state, auto-creating/loading if none exists.
    pub async fn get_state(&self) -> Result<GameState, GameError> {
        if let Some(state) = self.inner.lock().await.state.clone() {
            return Ok(state);
        }

        // Hold the init lock across the create/load decision; re-check once we
        // have it, since another task may have initialized while we waited.
        let _init = self.init_lock.lock().await;
        if let Some(state) = self.inner.lock().await.state.clone() {
            return Ok(state);
        }

        if sample_game_enabled() {
            self.create_sample_game().await?;
        } else {
            let most_recent = self.inner.lock().await.store.get_most_recent_game_id()?;
            if let Some(game_id) = most_recent {
                self.load_game(game_id).await?;
            } else {
                self.create_game("trouble_brewing").await?;
            }
        }
        self.state().await
    }

    pub async fn create_game(&self, script_name: &str) -> Result<GameState, GameError> {
        let mut inner = self.inner.lock().await;
        let game_id = Uuid::new_v4();
        let event = GameEvent {
            id: Uuid::new_v4(),
            game_id,
            sequence: 0,
            timestamp: Utc::now(),
            payload: EventPayload::GameCreated {
                script_name: script_name.to_string(),
            },
        };
        let initial = GameState::initial(game_id, "");
        let new_state = apply(&initial, &event);
        inner.store.append(&event)?;
        inner.state = Some(new_state.clone());
        Ok(new_state)
    }

    pub async fn load_game(&self, game_id: Uuid) -> Result<GameState, GameError> {
        let mut inner = self.inner.lock().await;
        let events = inner.store.get_events(game_id, None)?;
        if events.is_empty() {
            return Err(GameError::GameNotFound(game_id));
        }
        let state = replay(&events)?;
        inner.state = Some(state.clone());
        Ok(state)
    }

    pub async fn rewind(&self, to_version: i64) -> Result<GameState, GameError> {
        let mut inner = self.inner.lock().await;
        let current = inner.state.clone().ok_or(GameError::NoActiveGame)?;
        if to_version < 1 || to_version > current.version {
            return Err(GameError::InvalidVersion(format!(
                "Invalid version {to_version} (current: {})",
                current.version
            )));
        }
        inner
            .store
            .delete_after_sequence(current.game_id, to_version - 1)?;
        let events = inner.store.get_events(current.game_id, None)?;
        let state = replay(&events)?;
        inner.state = Some(state.clone());
        Ok(state)
    }

    pub async fn fork(&self, from_version: i64) -> Result<GameState, GameError> {
        let mut inner = self.inner.lock().await;
        let current = inner.state.clone().ok_or(GameError::NoActiveGame)?;
        if from_version < 1 || from_version > current.version {
            return Err(GameError::InvalidVersion(format!(
                "Invalid version {from_version} (current: {})",
                current.version
            )));
        }
        let new_game_id = inner.store.fork_game(current.game_id, from_version - 1)?;
        let events = inner.store.get_events(new_game_id, None)?;
        let state = replay(&events)?;
        inner.state = Some(state.clone());
        Ok(state)
    }

    pub async fn get_history(&self) -> Result<Vec<GameEvent>, GameError> {
        let inner = self.inner.lock().await;
        let current = inner.state.clone().ok_or(GameError::NoActiveGame)?;
        Ok(inner.store.get_events(current.game_id, None)?)
    }

    pub async fn list_games(&self) -> Result<Vec<Uuid>, GameError> {
        Ok(self.inner.lock().await.store.get_all_game_ids()?)
    }

    async fn create_sample_game(&self) -> Result<(), GameError> {
        self.create_game("trouble_brewing").await?;

        let roles = [
            "Imp",
            "Baron",
            "Poisoner",
            "Scarlet Woman", // evil
            "Recluse",
            "Librarian",
            "Empath",
            "Investigator", // good
            "Mayor",
            "Fortune Teller",
            "Slayer",
            "Monk",
            "Virgin", // good
        ];
        self.dispatch(EventPayload::RolesIncluded {
            names: roles.iter().map(|s| s.to_string()).collect(),
        })
        .await?;

        let players = [
            ("Ryan", "Baron", "evil"),
            ("Yash", "Virgin", "good"),
            ("Other Ryan", "Imp", "evil"),
            ("Other Yash", "Poisoner", "evil"),
            ("Yet Another Ryan", "Scarlet Woman", "evil"),
            ("Yet Another Yash", "Recluse", "good"),
            ("Even More Ryan", "Librarian", "good"),
            ("Even More Yash", "Empath", "good"),
            ("Even Even More Ryan", "Mayor", "good"),
            ("Even Even More Yash", "Fortune Teller", "good"),
            ("Claude", "Monk", "good"),
        ];
        for (name, role, alignment) in players {
            self.dispatch(EventPayload::PlayerAdded {
                player_name: name.to_string(),
                character_name: role.to_string(),
                alignment: alignment.to_string(),
            })
            .await?;
        }

        for (player, effect) in [
            ("Yash", "Drunk"),
            ("Yash", "No Ability"),
            ("Ryan", "Is The Demon"),
        ] {
            self.dispatch(EventPayload::StatusEffectAdded {
                player_name: player.to_string(),
                effect: effect.to_string(),
            })
            .await?;
        }

        self.dispatch(EventPayload::PlayerAliveSet {
            player_name: "Yash".to_string(),
            is_alive: false,
            cleared_effects: Vec::new(),
        })
        .await?;
        self.dispatch(EventPayload::DeadVoteUsedSet {
            player_name: "Yash".to_string(),
            has_used_dead_vote: true,
        })
        .await?;

        Ok(())
    }
}

fn sample_game_enabled() -> bool {
    std::env::var("SAMPLE_GAME")
        .map(|v| matches!(v.to_lowercase().as_str(), "1" | "true"))
        .unwrap_or(false)
}
